service/history/mutableState.go

```diff
@@ -211,6 +211,8 @@ type (
 		UpdateUserTimer(*persistence.TimerInfo) error
 		UpdateCurrentVersion(version int64, forceUpdate bool) error
 		UpdateWorkflowStateCloseStatus(state int, closeStatus int) error
+		IsEventReapplied(runID string, eventID int64, version int64) bool
+		UpdateReappliedEvent(runID string, eventID int64, version int64)
```

Comment:
i believe the dedup functionality can be made generic, i.e. supporting events reapplication & existing signal functionality

Reply:
plz also check with @vancexu

Reply:
Yes. dedup of signal external wf and this can merged